src/protocol/ContactRequestChannel.cpp

```diff
@@ -0,0 +1,305 @@
+        connect(this, &Channel::channelOpened, this, &Channel::closeChannel, Qt::QueuedConnection);
+    return true;
+}
+
+void ContactRequestChannel::setResponseStatus(Status status, const QString &message)
+{
+    if (m_responseStatus == status)
+        return;
+
+    if (direction() != Inbound) {
+        BUG() << "Can't set the response on an outbound contact request";
+        return;
+    }
+
+    using namespace Data::ContactRequest;
+    if (m_responseStatus > Response::Pending)
+        BUG() << "Response status is already a final state" << m_responseStatus << "but was changed to" << status;
+
+    m_responseStatus = status;
+    m_responseErrorMessage = message;
+
+    // If the channel is already open, the response is sent as a separate packet
+    if (isOpened()) {
+        Response response;
+        response.set_status(m_responseStatus);
+        if (!m_responseErrorMessage.isEmpty())
+            response.set_error_message(m_responseErrorMessage.toStdString());
+        sendMessage(response);
+
+        if (m_responseStatus > Response::Pending)
+            closeChannel();
+    }
+}
+
+bool ContactRequestChannel::allowOutboundChannelRequest(Data::Control::OpenChannel *request)
+{
+    if (connection()->direction() != Connection::ClientSide ||
+        connection()->purpose() != Connection::Purpose::OutboundRequest)
+    {
+        BUG() << "ContactRequestChannel can only be used on OutboundRequest connections. Has purpose"
+              << int(connection()->purpose());
+        return false;
+    }
+
+    if (connection()->findChannel<ContactRequestChannel>()) {
+        BUG() << "ContactRequestChannel can only be used once per connection";
+        return false;
+    }
+
+    QScopedPointer<Data::ContactRequest::ContactRequest> contactData(new Data::ContactRequest::ContactRequest);
+    if (!m_nickname.isEmpty())
+        contactData->set_nickname(m_nickname.toStdString());
+    if (!m_message.isEmpty())
+        contactData->set_message_text(m_message.toStdString());
+
+    request->SetAllocatedExtension(Data::ContactRequest::contact_request, contactData.take());
+    return true;
+}
+
+bool ContactRequestChannel::processChannelOpenResult(const Data::Control::ChannelResult *result)
+{
+    if (!result->HasExtension(Data::ContactRequest::response)) {
+        qDebug() << "Expected a response for the contact request";
+        return false;
+    }
+
+    Data::ContactRequest::Response response = result->GetExtension(Data::ContactRequest::response);
+    return handleResponse(&response);
+}
+
+void ContactRequestChannel::receivePacket(const QByteArray &packet)
+{
+    Data::ContactRequest::Response response;
+    if (!response.ParseFromArray(packet.constData(), packet.size())) {
+        closeChannel();
+        return;
+    }
+
+    if (!handleResponse(&response))
+        closeChannel();
+}
+
+bool ContactRequestChannel::handleResponse(const Data::ContactRequest::Response *response)
```

Comment:
Ah, a tricky trap!

The `handle*` functions are handling remote data. If we did `BUG()` on them, those 'bug's would be remotely triggerable. That's not dangerous for a release build, but it's not correct either.